### PR TITLE
BED-4463: Implement BOM encoding package and refactor WriteAndValidateJSON for improved Unicode handling

### PR DESCRIPTION
**Pull request by neumachen**
## Description

This PR introduces a robust package for handling Byte Order Mark (BOM) encoding detection and UTF conversions and refactors the `WriteAndValidateJSON` function to use this new package. The changes include:

1. Implement an `Encoding` interface to handle different encodings.
2. Precise BOM detection for UTF-8, UTF-16 (BE/LE), and UTF-32 (BE/LE).
3. Implementation of UTF-16 and UTF-32 to UTF-8 conversion with proper handling of surrogate pairs.
4. Refactoring of `WriteAndValidateJSON` to use the new `bomenc` package for BOM handling and UTF normalization.
5. Improved error handling and simplified logic in `WriteAndValidateJSON`.

## Motivation and Context

This PR addresses: [BED-4463](https://specterops.atlassian.net/browse/BED-4463)

These changes are necessary to enhance our management of diverse text encodings across the codebase, especially in IO operations involving potentially BOM-prefixed content. The introduction of the `bomenc` package offers a more robust and comprehensive solution for handling various Unicode encodings. The refactoring of `WriteAndValidateJSON` leverages this package to simplify the function and enhance its versatility and reliability.

## How Has This Been Tested?

Developed comprehensive test suites for the `bomenc` package, including:

- Tests for encoding detection, conversion, and edge cases.
- Tests for custom encodings and the new `HasSequence` method.
- Tests covering various scenarios of BOM detection and removal.
- Tests for UTF normalization with different input encodings.

Testing the refactored `WriteAndValidateJSON` function to ensure correct behavior with inputs of various encodings and with/without BOMs.

## Types of changes

- New feature (non-breaking change which adds functionality)
- Chore (a change that does not modify the application functionality)

## Checklist:

- [x] I have met the contributing prerequisites
  - Assigned myself to this PR
  - Added the appropriate labels
  - Associated an issue: https://github.com/SpecterOps/BloodHound/issues/672
  - Read the Contributing guide: https://github.com/SpecterOps/BloodHound/wiki/Contributing
- [x] I have ensured that related documentation is up-to-date
  - Open API docs
  - Code comments (GoDocs / JSDocs)
- [x] I have followed proper test practices
  - Added/updated tests to cover my changes
  - All new and existing tests passed

[BED-4463]: https://specterops.atlassian.net/browse/BED-4463?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by neumachen** [comment]
I have read the CLA Document and I hereby sign the CLA

File: packages/go/bomenc/encodings.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+// UTF16LE represents the UTF-16 Little Endian encoding.
+var UTF16LE Encoding = &bomEncoding{
+	encodingType:    "UTF-16 LE",
+	sequence:        []byte{0xFF, 0xFE}, // UTF-16 LE BOM sequence
+	hasSequenceFunc: isUTF16LE,
+}
+
+// UTF32BE represents the UTF-32 Big Endian encoding.
+var UTF32BE Encoding = &bomEncoding{
```
Comment (zinic) [inline packages/go/bomenc/encodings.go:133]: Minor nit that these could all be value structs instead of pointers since you're not mutating the `bomEncoding` struct after the fact.

File: packages/go/bomenc/encodings.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+type bomEncoding struct {
+	encodingType    string                 // A human-readable name for the encoding
+	sequence        []byte                 // The BOM sequence for this encoding
+	hasSequenceFunc func(data []byte) bool // Function to check if data starts with this encoding's BOM
+}
+
+// String returns the human-readable name of the encoding.
+// This method fulfills the Encoding interface and provides a simple way
+// to get a string representation of the encoding.
+func (b bomEncoding) String() string {
```
Comment (zinic) [inline packages/go/bomenc/encodings.go:54]: As per our golang standards all receivers should be `s`. I see this in a few places but will only write this comment for it.

File: packages/go/bomenc/normalize.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+	"bytes"
+	"errors"
+	"io"
+	"unicode/utf8"
+)
+
+// Normalizer interface defines the methods for accessing normalized content.
+// This allows consumers of the package to work with normalized data without
+// knowing the specifics of how the normalization was performed.
+type Normalizer interface {
```
Comment (zinic) [inline packages/go/bomenc/normalize.go:29]: If there's only one implementation do we need this interface?

File: packages/go/bomenc/normalize.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	}
+
+	return NormalizeBytesToUTF8(data, detectedBOMEncoding)
+}
+
+// NormalizeBytesToUTF8 converts the given bytes to UTF-8 based on the specified encoding.
+// This function is the core of the normalization process, handling different encodings
+// and converting them to UTF-8.
+func NormalizeBytesToUTF8(data []byte, enc Encoding) (Normalizer, error) {
+	var content []byte
```
Comment (zinic) [inline packages/go/bomenc/normalize.go:140]: Collapse var statements:

```
var a string
var b string
```

to

```
var (
   a string
   b string
)
```

File: packages/go/bomenc/utf16.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+//     big-endian and little-endian encodings with minimal code duplication.
+//
+//  5. Performance in loops: When processing large amounts of text, the
+//     performance benefits of bitwise operations become significant due
+//     to the number of times these operations are repeated.
+//
+//  6. Low-level control: Bitwise operations provide fine-grained control
+//     over individual bits, which is necessary for correct interpretation
+//     of multi-byte character encodings.
+func utf16ToUTF8(data []byte, bigEndian bool) ([]byte, error) {
```
Comment (zinic) [inline packages/go/bomenc/utf16.go:59]: There are a lot of bitwise operations and other arcane components in this function. Is there truly no upstream stdlib or library function that does the same so we don't need to maintain this?

See: https://pkg.go.dev/unicode/utf16

File: packages/go/bomenc/utf32.go [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+//     beneficial for memory usage, especially when processing large files.
+//
+//  6. Portability: Bitwise operations behave consistently across different
+//     hardware architectures, ensuring our code works reliably on various systems.
+//
+//  7. Educational value: Understanding and using bitwise operations provides
+//     insights into low-level data representation, which is valuable knowledge
+//     for any programmer working with different character encodings or
+//     binary protocols.
+func utf32ToUTF8(data []byte, bigEndian bool) ([]byte, error) {
```
Comment (zinic) [inline packages/go/bomenc/utf32.go:56]: There are a lot of bitwise operations and other arcane components in this function. Is there truly no upstream stdlib or library function that does the same so we don't need to maintain this?

See: https://pkg.go.dev/golang.org/x/text/encoding/unicode/utf32

**Review by zinic — CHANGES_REQUESTED** [review]
Great effort, lots of good comments and the test cases are sound. I have some concerns around adopting some of the implementation here and request the author clear some additional nits.

File: packages/go/bomenc/normalize.go (hunk 3 above)
Comment (neumachen) [reply packages/go/bomenc/normalize.go:29]: @zinic not necessarily, but I wanted to convey that the function returning this interface is meant read-only.

**Comment by neumachen** [comment]
@zinic here's my finding for the alternative:

```go
func utf16ToUTF8Alternative(data []byte, bigEndian bool) ([]byte, error) {
	var u16s []uint16
	for i := 0; i < len(data); i += 2 {
		if bigEndian {
			u16s = append(u16s, binary.BigEndian.Uint16(data[i:i+2]))
		} else {
			u16s = append(u16s, binary.LittleEndian.Uint16(data[i:i+2]))
		}
	}

	runes := utf16.Decode(u16s)
	var buf []byte
	for _, r := range runes {
		b := make([]byte, utf8.RuneLen(r))
		utf8.EncodeRune(b, r)
		buf = append(buf, b...)
	}
	return buf, nil
}

func utf32ToUTF8Alternative(data []byte, bigEndian bool) ([]byte, error) {
	var buf []byte
	for i := 0; i < len(data); i += 4 {
		var r rune
		if bigEndian {
			r = rune(binary.BigEndian.Uint32(data[i : i+4]))
		} else {
			r = rune(binary.LittleEndian.Uint32(data[i : i+4]))
		}

		// Check if the rune is a valid Unicode code point
		if r > unicode.MaxRune || (r >= 0xD800 && r <= 0xDFFF) {
			r = unicode.ReplacementChar
		}

		b := make([]byte, utf8.RuneLen(r))
		utf8.EncodeRune(b, r)
		buf = append(buf, b...)
	}
	return buf, nil
}
```

[bench_1.txt](https://github.com/user-attachments/files/16056676/bench_1.txt)
[bench_2.txt](https://github.com/user-attachments/files/16056677/bench_2.txt)
[bench_3.txt](https://github.com/user-attachments/files/16056678/bench_3.txt)
[bench_4.txt](https://github.com/user-attachments/files/16056679/bench_4.txt)
[bench_5.txt](https://github.com/user-attachments/files/16056680/bench_5.txt)


```shell
goos: darwin
goarch: arm64
pkg: github.com/specterops/bloodhound/boomenc
                         │ bench_1.txt  │            bench_2.txt             │            bench_3.txt             │            bench_4.txt             │            bench_5.txt             │
                         │    sec/op    │    sec/op     vs base              │    sec/op     vs base              │    sec/op     vs base              │    sec/op     vs base              │
UTF16ToUTF8                3.297µ ± ∞ ¹   3.214µ ± ∞ ¹       ~ (p=0.056 n=5)   3.200µ ± ∞ ¹  -2.94% (p=0.008 n=5)   3.211µ ± ∞ ¹       ~ (p=0.087 n=5)   3.261µ ± ∞ ¹       ~ (p=0.151 n=5)
UTF16ToUTF8-2              2.420µ ± ∞ ¹   2.429µ ± ∞ ¹       ~ (p=0.802 n=5)   2.419µ ± ∞ ¹       ~ (p=0.540 n=5)   2.417µ ± ∞ ¹       ~ (p=0.810 n=5)   2.433µ ± ∞ ¹       ~ (p=0.206 n=5)
UTF16ToUTF8-4              2.409µ ± ∞ ¹   2.402µ ± ∞ ¹       ~ (p=0.071 n=5)   2.413µ ± ∞ ¹       ~ (p=0.246 n=5)   2.405µ ± ∞ ¹       ~ (p=0.302 n=5)   2.407µ ± ∞ ¹       ~ (p=0.667 n=5)
UTF32ToUTF8                3.031µ ± ∞ ¹   3.049µ ± ∞ ¹  +0.59% (p=0.016 n=5)   3.012µ ± ∞ ¹  -0.63% (p=0.008 n=5)   3.041µ ± ∞ ¹       ~ (p=0.063 n=5)   3.017µ ± ∞ ¹  -0.46% (p=0.008 n=5)
UTF32ToUTF8-2              2.241µ ± ∞ ¹   2.251µ ± ∞ ¹  +0.45% (p=0.032 n=5)   2.248µ ± ∞ ¹       ~ (p=0.159 n=5)   2.247µ ± ∞ ¹       ~ (p=0.437 n=5)   2.244µ ± ∞ ¹       ~ (p=0.881 n=5)
UTF32ToUTF8-4              2.229µ ± ∞ ¹   2.226µ ± ∞ ¹       ~ (p=0.587 n=5)   2.227µ ± ∞ ¹       ~ (p=0.667 n=5)   2.226µ ± ∞ ¹       ~ (p=0.690 n=5)   2.228µ ± ∞ ¹       ~ (p=0.849 n=5)
UTF16ToUTF8Alternative     7.600µ ± ∞ ¹   7.544µ ± ∞ ¹       ~ (p=0.421 n=5)   7.538µ ± ∞ ¹       ~ (p=0.460 n=5)   7.498µ ± ∞ ¹       ~ (p=0.095 n=5)   7.587µ ± ∞ ¹       ~ (p=0.841 n=5)
UTF16ToUTF8Alternative-2   7.194µ ± ∞ ¹   7.148µ ± ∞ ¹       ~ (p=0.095 n=5)   7.161µ ± ∞ ¹       ~ (p=0.238 n=5)   7.143µ ± ∞ ¹       ~ (p=0.206 n=5)   7.209µ ± ∞ ¹       ~ (p=1.000 n=5)
UTF16ToUTF8Alternative-4   7.151µ ± ∞ ¹   7.170µ ± ∞ ¹       ~ (p=0.841 n=5)   7.180µ ± ∞ ¹       ~ (p=0.310 n=5)   7.149µ ± ∞ ¹       ~ (p=0.730 n=5)   7.166µ ± ∞ ¹       ~ (p=0.548 n=5)
UTF32ToUTF8Alternative     1.058µ ± ∞ ¹   1.037µ ± ∞ ¹  -1.98% (p=0.008 n=5)   1.047µ ± ∞ ¹       ~ (p=0.127 n=5)   1.039µ ± ∞ ¹  -1.80% (p=0.008 n=5)   1.035µ ± ∞ ¹  -2.17% (p=0.008 n=5)
UTF32ToUTF8Alternative-2   1.020µ ± ∞ ¹   1.022µ ± ∞ ¹       ~ (p=0.119 n=5)   1.031µ ± ∞ ¹       ~ (p=0.135 n=5)   1.018µ ± ∞ ¹       ~ (p=0.389 n=5)   1.018µ ± ∞ ¹       ~ (p=0.095 n=5)
UTF32ToUTF8Alternative-4   1.018µ ± ∞ ¹   1.022µ ± ∞ ¹  +0.39% (p=0.008 n=5)   1.023µ ± ∞ ¹       ~ (p=0.095 n=5)   1.019µ ± ∞ ¹       ~ (p=0.317 n=5)   1.019µ ± ∞ ¹       ~ (p=0.619 n=5)
geomean                    2.659µ         2.650µ        -0.34%                 2.651µ        -0.30%                 2.645µ        -0.54%                 2.652µ        -0.26%
¹ need >= 6 samples for confidence interval at level 0.95

                         │  bench_1.txt  │              bench_2.txt              │              bench_3.txt              │              bench_4.txt              │              bench_5.txt              │
                         │     B/op      │     B/op       vs base                │     B/op       vs base                │     B/op       vs base                │     B/op       vs base                │
UTF16ToUTF8                3.938Ki ± ∞ ¹   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF16ToUTF8-2              3.938Ki ± ∞ ¹   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF16ToUTF8-4              3.938Ki ± ∞ ¹   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF32ToUTF8                3.938Ki ± ∞ ¹   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF32ToUTF8-2              3.938Ki ± ∞ ¹   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF32ToUTF8-4              3.938Ki ± ∞ ¹   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   3.938Ki ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF16ToUTF8Alternative     12.28Ki ± ∞ ¹   12.28Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   12.28Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   12.28Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   12.28Ki ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF16ToUTF8Alternative-2   12.28Ki ± ∞ ¹   12.28Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   12.28Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   12.28Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   12.28Ki ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF16ToUTF8Alternative-4   12.28Ki ± ∞ ¹   12.28Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   12.28Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   12.28Ki ± ∞ ¹       ~ (p=1.000 n=5) ²   12.28Ki ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF32ToUTF8Alternative       352.0 ± ∞ ¹     352.0 ± ∞ ¹       ~ (p=1.000 n=5) ²     352.0 ± ∞ ¹       ~ (p=1.000 n=5) ²     352.0 ± ∞ ¹       ~ (p=1.000 n=5) ²     352.0 ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF32ToUTF8Alternative-2     352.0 ± ∞ ¹     352.0 ± ∞ ¹       ~ (p=1.000 n=5) ²     352.0 ± ∞ ¹       ~ (p=1.000 n=5) ²     352.0 ± ∞ ¹       ~ (p=1.000 n=5) ²     352.0 ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF32ToUTF8Alternative-4     352.0 ± ∞ ¹     352.0 ± ∞ ¹       ~ (p=1.000 n=5) ²     352.0 ± ∞ ¹       ~ (p=1.000 n=5) ²     352.0 ± ∞ ¹       ~ (p=1.000 n=5) ²     352.0 ± ∞ ¹       ~ (p=1.000 n=5) ²
geomean                    2.844Ki         2.844Ki        +0.00%                   2.844Ki        +0.00%                   2.844Ki        +0.00%                   2.844Ki        +0.00%
¹ need >= 6 samples for confidence interval at level 0.95
² all samples are equal

                         │ bench_1.txt │             bench_2.txt             │             bench_3.txt             │             bench_4.txt             │             bench_5.txt             │
                         │  allocs/op  │  allocs/op   vs base                │  allocs/op   vs base                │  allocs/op   vs base                │  allocs/op   vs base                │
UTF16ToUTF8                6.000 ± ∞ ¹   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF16ToUTF8-2              6.000 ± ∞ ¹   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF16ToUTF8-4              6.000 ± ∞ ¹   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF32ToUTF8                6.000 ± ∞ ¹   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF32ToUTF8-2              6.000 ± ∞ ¹   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF32ToUTF8-4              6.000 ± ∞ ¹   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²   6.000 ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF16ToUTF8Alternative     517.0 ± ∞ ¹   517.0 ± ∞ ¹       ~ (p=1.000 n=5) ²   517.0 ± ∞ ¹       ~ (p=1.000 n=5) ²   517.0 ± ∞ ¹       ~ (p=1.000 n=5) ²   517.0 ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF16ToUTF8Alternative-2   517.0 ± ∞ ¹   517.0 ± ∞ ¹       ~ (p=1.000 n=5) ²   517.0 ± ∞ ¹       ~ (p=1.000 n=5) ²   517.0 ± ∞ ¹       ~ (p=1.000 n=5) ²   517.0 ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF16ToUTF8Alternative-4   517.0 ± ∞ ¹   517.0 ± ∞ ¹       ~ (p=1.000 n=5) ²   517.0 ± ∞ ¹       ~ (p=1.000 n=5) ²   517.0 ± ∞ ¹       ~ (p=1.000 n=5) ²   517.0 ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF32ToUTF8Alternative     105.0 ± ∞ ¹   105.0 ± ∞ ¹       ~ (p=1.000 n=5) ²   105.0 ± ∞ ¹       ~ (p=1.000 n=5) ²   105.0 ± ∞ ¹       ~ (p=1.000 n=5) ²   105.0 ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF32ToUTF8Alternative-2   105.0 ± ∞ ¹   105.0 ± ∞ ¹       ~ (p=1.000 n=5) ²   105.0 ± ∞ ¹       ~ (p=1.000 n=5) ²   105.0 ± ∞ ¹       ~ (p=1.000 n=5) ²   105.0 ± ∞ ¹       ~ (p=1.000 n=5) ²
UTF32ToUTF8Alternative-4   105.0 ± ∞ ¹   105.0 ± ∞ ¹       ~ (p=1.000 n=5) ²   105.0 ± ∞ ¹       ~ (p=1.000 n=5) ²   105.0 ± ∞ ¹       ~ (p=1.000 n=5) ²   105.0 ± ∞ ¹       ~ (p=1.000 n=5) ²
geomean                    37.39         37.39        +0.00%                   37.39        +0.00%                   37.39        +0.00%                   37.39        +0.00%
¹ need >= 6 samples for confidence interval at level 0.95
² all samples are equal
```

## Benchmark Results

1. UTF-16 to UTF-8 Conversion:
   Original implementation (UTF16ToUTF8):
   - Average time: ~3.2-3.3 µs/op
   - Memory allocation: 4032 B/op
   - Allocations: 6 allocs/op

   Alternative implementation (UTF16ToUTF8Alternative):
   - Average time: ~7.5-7.6 µs/op
   - Memory allocation: 12576 B/op
   - Allocations: 517 allocs/op

   Analysis: The original implementation is significantly faster (more than 2x) and uses less memory than the alternative. It also has far fewer allocations, which contributes to its better performance.

2. UTF-32 to UTF-8 Conversion:
   Original implementation (UTF32ToUTF8):
   - Average time: ~3.0-3.1 µs/op
   - Memory allocation: 4032 B/op
   - Allocations: 6 allocs/op

   Alternative implementation (UTF32ToUTF8Alternative):
   - Average time: ~1.0-1.1 µs/op
   - Memory allocation: 352 B/op
   - Allocations: 105 allocs/op

   Analysis: Interestingly, the alternative implementation for UTF-32 to UTF-8 is faster (about 3x) than the original. It also uses less memory but with more allocations.

3. CPU Core Scaling:
  In general, when running tests with 1, 2, and 4 CPU cores, performance needs to improve more when increasing the number of cores, suggesting that these operations are not highly parallelizable or that the overhead of parallelization outweighs the benefits for these specific tasks.
4. Consistency:
   The results are consistent across the five runs, with only slight variations in timing suggesting that the benchmarks are reliable and the performance characteristics are stable.
5. Memory and Allocations:
   For each implementation, the memory usage and number of allocations remain constant across different runs and core counts.

Conclusions:
1. For UTF-16 to UTF-8 conversion, the original implementation is superior in speed and resource usage.
2. For UTF-32 to UTF-8 conversion, the alternative implementation is faster and uses less memory, making it the better choice despite having more allocations.
3. These operations benefit only a little from multiple CPU cores, at least not in their current implementation.
4. The original implementations perform consistently across UTF-16 and UTF-32 conversions, while the alternative implementations have very different characteristics between UTF-16 and UTF-32.

Recommendations:
1. Stick with the original implementation for UTF-16 to UTF-8 conversion.
2. Consider using the alternative implementation for UTF-32 to UTF-8 conversion, but profile in real-world scenarios to ensure the increased number of allocations doesn't cause issues (e.g., increased GC pressure).
3. If you need to optimize further, focus on reducing allocations, especially in the alternative implementations.
4. Given the lack of improvement with multiple cores, don't prioritize parallelization for these specific operations.

These benchmarks provide valuable insights into the performance characteristics of your UTF conversion implementations, allowing you to make informed decisions about which methods to use in your application.

## Original Implementations (using bitwise operations):

### Pros:
1. Compact code: The implementations are generally more concise.
2. Direct control: Provides fine-grained control over the conversion process.
3. No dependencies: Don't rely on additional standard library packages beyond the basics.
4. Consistent approach: Using similar logic for UTF-16 and UTF-32 conversions.

### Cons:
1. Readability: Bitwise operations can be more complex to read and understand, especially for developers who need to become more familiar with low-level operations.
2. Error-prone: Bit manipulation is susceptible to subtle errors that can be difficult to spot.
3. Limited built-in safety: Lacks some built-in safety checks with standard library functions.
4. Harder to modify: Log changes might require careful consideration of the bit-level operations.

## Alternative Implementations (using standard library functions):

### Pros:
1. Readability: Generally easier to understand at a glance, especially for developers familiar with Go's standard library.
2. Built-in safety: Standard library functions often include built-in checks and safeguards.
3. Standardization: Uses well-tested and widely-used standard library functions.
4. Easier to modify: Logic changes can often be made by adjusting function calls rather than rewriting bit manipulations.

### Cons:
1. Less control: It relies on the behavior of standard library functions, which may only sometimes perfectly align with specific needs.
2. Potential overhead: This may include unnecessary checks or operations for the specific use case.
3. More verbose: This can result in more extended code, especially when chaining multiple standard library function calls.
4. Dependency on the standard library: Changes to the standard library in future Go versions could affect the code.

## General Observations:
1. Consistency: The original implementations have a more consistent approach between UTF-16 and UTF-32, while the alternative implementations differ more significantly.
2. Learning curve: The original implementations require a deeper understanding of Unicode and bitwise operations, while the alternative implementations rely more on knowledge of Go's standard library.
3. Documentation needs: The original implementations likely require more detailed comments explaining the bitwise logic, while the alternative implementations might be more self-documenting.
4. Testing: The alternative implementations might be more straightforward to test in isolation, as they rely on standard library functions that are already well-tested.
5. Flexibility: The original implementations might be more flexible for handling edge cases or specific optimizations, while the alternative implementations are more rigid but potentially more robust for general cases.

Regarding maintainability, these approaches often depend on the team's expertise, the project's specific requirements, and the expected frequency and nature of future modifications. The alternative implementations might be more maintainable in a general sense, especially for teams more familiar with Go's standard library. However, the original implementations could be more maintainable in the long run for teams with low-level programming expertise or projects with particular performance requirements.

File: packages/go/bomenc/normalize.go (hunk 3 above)
Comment (zinic) [reply packages/go/bomenc/normalize.go:29]: :+1:

**Review by zinic — APPROVED** [review]
Approved with the expectation that we'll move to the alternative version that uses the Golang stdlib in a follow up PR. Thanks for the great legwork here!
